### ISSUE TITLE
[Feature]: Should be able to add snapshots

### ISSUE THREAD

**Issue by uhidontkno** [dockur/windows#485]
### Is there an existing feature request for this?

- [X] I have searched the existing feature requests

### Is your proposal related to a problem?

Instead of waiting up to 15 minutes for Windows to install, I propose that a snapshot feature should be implemented so that there should be a way to install windows once, make a snapshot for it, then be able to load it anytime for instant Windows installations.

### Describe the solution you'd like.

One solution could be to have a way to save the entire VM's disk to somewhere else using `docker cp` for example, then feeding that file into the Windows Docker Container, or simply implement [QEMU snapshots support](https://wiki.qemu.org/Features/Snapshots)

### Describe alternatives you've considered.

I haven't tried anything yet.

**Comment by lucaspdude** [comment]
Interested in this also. 

At work we use some old VB6 configured machine and it would be really nice if we could make snapshots since most of the configuration is manual for this environment.

**Comment by baiomys** [comment]
ZFS volume can be passed to docker instead of /storage via 
/dev/zvol/mypool/myvol:/disk1
It has native support for compression, snapshots, encryption, etc.
So it is worth spending couple of days reading ZFS for dummies.
=)

**Comment by TeroKeso** [comment]
ZFS passthrough would be a good addition to readme.md. Like @baiomys did say just link them to some instructions page like https://blog.victormendonca.com/2020/11/03/zfs-for-dummies/ if they need more info

**Comment by TeroKeso** [comment]
You are absolutely right about the difficulty level. I just understand what I'm doing and that's why I suggested writing the instructions. I'm not experienced enough to write instructions and see all possibilities. 

Zvol or similar helps with snapshot+backup when `/storage` is just one big file (HDD). 

Thank you for taking the time to reply and I'm sorry that my question got a bit out of the title/topic

**Comment by baiomys** [comment]
ZFS makes you think completely different, it extends old plain storage paradigm in new dimension. 
That's why learning curve is kind of steep.

**Comment by abbbi** [comment]
hi, 

note that you would have to generalize your windows installation using sysprep if you want to clone
an run the image multiple times within the same network...

My idea would probably to docker compose up with an custom install.bat, that does an unattended
sysprep of the system, and then use this sysprepped image for the snapshots.

**Comment by abbbi** [comment]
this is my take on this:

https://github.com/abbbi/docker-windows-sysprep

**Comment by inetic** [comment]
> So for backing up or restoring machines, all you need to do is make a copy of `/example/folder`. And if you have multiple folders, you can easily switch between them by changing that line in your compose file.
> 
> ... The only difference is that you cannot make a copy of `/storage` while the machine is powered on ....

I presume this would apply also when making snapshots with ZFS or BTRFS? That is, does one need to power the machine off to make a snapshot?

It would be quite a win if that wasn't the case.

If anyone here has more experiences with those file systems and how taking snapshots on a running Windows VM would affect it, please let us know.

**Comment by kroese** [comment]
It can be done pretty easily in another way.

The first step is to convert the machine to QCOW2 instead of the default (raw) disk format:

```yaml
environment:
  DISK_FMT: "qcow2"
```

Then you can use the snapshot functionality QCOW2 provides, like so:

**Create a snapshot:**

```qemu-img snapshot -c snapshot_name data.qcow2```

**Restore a snapshot:**

```qemu-img snapshot -a snapshot_name data.qcow2```

**Delete a snapshot:**

```qemu-img snapshot -d snapshot_name data.qcow2```

**List all snapshots:**

```qemu-img snapshot -l data.qcow2```

Now, to perform these same operations on a running machine you can connect to the QEMU monitor port:

```yaml
environment:
  MONITOR: "telnet:0.0.0.0:7777,server,nowait,nodelay"
  HOST_PORTS: "7777"
```

If you then connect to port 7777 of the container via telnet, you can send these commands:

**Create a snapshot of the whole virtual machine:**
`savevm <tag>`

**Restore the whole virtual machine to a snapshot:**
`loadvm <tag>`

**Delete a snapshot:**
`delvm <tag>`

This uses the native functionality QEMU has builtin for this purpose. But I am not sure this more complicated way is even needed. There is a good chance that just for creating snapshots you can use `qemu-img` on a running machine, which is much easier. Only for restoring a snapshot to a running machine you will need this telnet-method.
